nssa/program_methods/guest/src/bin/amm.rs

```diff
@@ -0,0 +1,3356 @@
+    if pre_states.len() != 7 {
+        panic!("Invalid number of input accounts")
+    }
+
+    if balance_in.len() != 2 {
+        panic!("Invalid number of input balances")
+    }
+
+    let pool = &pre_states[0];
+    let vault_a = &pre_states[1];
+    let vault_b = &pre_states[2];
+    let pool_lp = &pre_states[3];
+    let user_holding_a = &pre_states[4];
+    let user_holding_b = &pre_states[5];
+    let user_holding_lp = &pre_states[6];
+
+    let amount_a = balance_in[0];
+    let amount_b = balance_in[1];
+
+    // Prevents pool constant coefficient (k) from being 0.
+    if amount_a == 0 || amount_b == 0 {
+        panic!("Balances must be nonzero")
+    }
+
+    // Verify token_a and token_b are different
+    let definition_token_a_id = TokenHolding::parse(&user_holding_a.account.data)
+        .expect("New definition: AMM Program expects valid Token Holding account for Token A").definition_id;
+    let definition_token_b_id = TokenHolding::parse(&user_holding_b.account.data)
+        .expect("New definition: AMM Program expects valid Token Holding account for Token B").definition_id;
+ 
+    // both instances of the same token program
+    let token_program = user_holding_a.account.program_owner;
+
+    if definition_token_a_id == definition_token_b_id {
+        panic!("Cannot set up a swap for a token with itself")
+    }
+
+    if pool.account_id != compute_pool_pda(amm_program_id.clone(),
+                                            definition_token_a_id.clone(),
+                                            definition_token_b_id.clone()) {
+        panic!("Pool Definition Account ID does not match PDA");
+    }
+
+    if vault_a.account_id != compute_vault_pda(amm_program_id.clone(),
+                                                pool.account_id.clone(),
+                                                definition_token_a_id.clone()) ||
+        vault_b.account_id != compute_vault_pda(amm_program_id.clone(),
+                                                pool.account_id.clone(),
+                                                definition_token_b_id.clone()) {
+        panic!("Vault ID does not match PDA");        
+    }
+
+    if pool_lp.account_id != compute_liquidity_token_pda(amm_program_id.clone(),
+                                                        pool.account_id.clone()) {
+        panic!("Liquidity pool Token Definition Account ID does not match PDA");
+    }
+
+    // Verify that Pool Account is not active
+    let pool_account_data = if pool.account == Account::default() {
+        PoolDefinition::default()
+    } else { 
+        PoolDefinition::parse(&pool.account.data).expect("AMM program expects a valid Pool account")
+    };
+
+    if pool_account_data.active {
+        panic!("Cannot initialize an active Pool Definition")
+    }
+
+    //3. LP Token minting calculation
+    // We assume LP is based on the initial deposit amount for Token_A.
+
+    // 5. Update pool account
```

Comment:
This 3 and 5 are the only step numbers in this function. Probably forget to remove them